packages/peregrine/lib/store/actions/cart/asyncActions.js

```diff
@@ -43,7 +43,10 @@ export const createCart = payload =>
 
             dispatch(actions.getCart.receive(receivePayload));
         } catch (error) {
+            // If we are unable to create a cart, the cart can't function, so
+            // we forcibly throw so the upstream actions won't retry.
             dispatch(actions.getCart.receive(error));
+            throw new Error('Unable to create cart');
```

Comment:
This feels a bit rough but it feels necessary. The retry logic in other actions necessitate that we somehow inform them of an unrecoverable error. I have chosen to throw in the root action and catch upstream but I wonder if there is a better way to do it. Perhaps propagate the error to the store and then, if an error such as that is present at the start of an action, bail?

```js
// reducers/cart.js
const reducerMap = {
    [actions.getCart.receive]: (state, { payload, error }) => {
        if (error) {
            return {
                ...initialState,
                getCartError: payload
            };
        }

        return {
            ...state,
            cartId: String(payload)
        };
    },
    ...
}

// cart/asyncActions.js
export const getCartDetails = payload => {
    return async function thunk(dispatch, getState) {
        const { cart, user } = getState();
        const { getCartError } = cart;

        if (getCartError) {
            // unrecoverable!
            return;
        }
        ...
    }
}

```

Reply:
No, this looks fine. I think it's standard practice to notify the reducer of an error, so it can be added to state, then follow up by throwing the error (or rejecting with it, if we were using promises directly).

Reply:
Ok - for now I've added the cart error to the state slice even though we won't do anything outright with it.

Reply:
It does seem preferable for these actions to check for and create a cart before attempting their own work instead of issuing a network request we know is going to fail and then doing the create + retry in a `catch`.

We could save ourselves processing time and network requests when the cart is bad.

And I don't think what @jimbo is saying contradicts that. This `throw` in `createCart` should stay, and also all the other actions could check for the presence of `getCartError` in state before attempting their work.